src/Course/Monad.hs

```diff
@@ -32,48 +29,39 @@ infixr 1 =<<
 -- >>> (\x -> ExactlyOne(x+1)) =<< ExactlyOne 2
 -- ExactlyOne 3
 instance Monad ExactlyOne where
-  (=<<) ::
-    (a -> ExactlyOne b)
-    -> ExactlyOne a
-    -> ExactlyOne b
-  (=<<) =
-    error "todo: Course.Monad (=<<)#instance ExactlyOne"
+  (=<<) :: (a -> ExactlyOne b) -> ExactlyOne a -> ExactlyOne b
+  f =<< (ExactlyOne x) = f x
+
+-- Left identity: return a >>= f ≡ f a
 
+-- Right identity: m >>= return ≡ m
+
+-- Associativity: (m >>= f) >>= g ≡ m >>= (\x -> f x >>= g)
 -- | Binds a function on a List.
 --
 -- >>> (\n -> n :. n :. Nil) =<< (1 :. 2 :. 3 :. Nil)
 -- [1,1,2,2,3,3]
 instance Monad List where
-  (=<<) ::
-    (a -> List b)
-    -> List a
-    -> List b
-  (=<<) =
-    error "todo: Course.Monad (=<<)#instance List"
+  (=<<) :: (a -> List b) -> List a -> List b
+  f =<< xs = flatten . map f $ xs
 
 -- | Binds a function on an Optional.
 --
 -- >>> (\n -> Full (n + n)) =<< Full 7
 -- Full 14
 instance Monad Optional where
-  (=<<) ::
-    (a -> Optional b)
-    -> Optional a
-    -> Optional b
-  (=<<) =
-    error "todo: Course.Monad (=<<)#instance Optional"
+  (=<<) :: (a -> Optional b) -> Optional a -> Optional b
+  _ =<< Empty = Empty
+  f =<< (Full x) = f x
 
 -- | Binds a function on the reader ((->) t).
 --
 -- >>> ((*) =<< (+10)) 7
 -- 119
 instance Monad ((->) t) where
-  (=<<) ::
-    (a -> ((->) t b))
-    -> ((->) t a)
-    -> ((->) t b)
-  (=<<) =
-    error "todo: Course.Monad (=<<)#instance ((->) t)"
+  -- (=<<) :: (a -> (t -> b)) -> (t -> a) -> (t -> b)
+  (=<<) :: (a -> ((->) t b)) -> ((->) t a) -> ((->) t b)
+  (f =<< g) x = f (g x) x
```

Comment:
nice! this took me ages to realise how to do this the first time someone explained it to me... :)

Reply:
haha, to be fair I wrote a few proofs of the reader monad in the past, I remember it took me some time to understand it too